### DISCUSSION TITLE
Parallel Chains Problem After Updating

### DISCUSSION THREAD

**Discussion by JoeSu112** [lnccbrown/HSSM#398] [Q&A] [answered]
Hi all,
I recently updated HSSM to the newest version (0.2.1). However, while I run the program that was able to run in the old version (0.2.0b1), some problem occurs:
```python
infer_DDM = DDM.sample(
    sampler="nuts_numpyro",  # type of sampler to choose,
    cores=2,  # how many cores to use
    chains=chains,  # how many chains to run
    draws=draws,  # number of draws from the markov chain
    tune=int(draws*0.2),  # number of burn-in samples
    idata_kwargs=dict(log_likelihood=False)
```
```
UserWarning: There are not enough devices to run parallel chains: expected 4 but got 2. 
Chains will be drawn sequentially. If you are running MCMC in CPU, 
consider using `numpyro.set_host_device_count(4)` at the beginning of your program. 
You can double-check how many devices are available in your system using `jax.local_device_count()`.
``` 
I'm sure that I could run parallel chains in the past versions. So I'm wondering how I can solve this problem and why this did not happen in the past versions.
Thanks for any help!!!

**Comment by JoeSu112** [comment]
Updated:
When I do the sampling again (on the same node), the above warning did not appear! It became as followed:
```
CUDA backend failed to initialize: Unable to load CUDA. Is it installed? (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
/home/miniconda3/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:66: UserWarning: 
Explicitly requested dtype float64 requested in astype is not available, and will be truncated to dtype float32. 
To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. 
See https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype)
/home/miniconda3/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:66: UserWarning: 
Explicitly requested dtype float64 requested in astype is not available, and will be truncated to dtype float32. 
To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. 
See https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype)
/home/b08303030/miniconda3/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:66: UserWarning: 
Explicitly requested dtype float64 requested in astype is not available, and will be truncated to dtype float32. 
To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. 
See https://github.com/google/jax#current-gotchas for more.
  return lax_numpy.astype(arr, dtype)
Compiling.. :   0%|                                        | 0/30 [00:00<?, ?it/s
/home/b08303030/miniconda3/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:66: UserWarning: 
Explicitly requested dtype float64 requested in astype is not available, and will be truncated to dtype float32. 
To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. 
See https://github.com/google/jax#current-gotchas for more.
``` 

And it can run parallel chains again!  However, the estimation time become a lot larger than expected.
In the past version, it took about 3-4 hours to estimate, but in this it takes more than a day.
It makes me really confused.   I'm wondering what the problem is.
Thanks for any help!!

**Comment by digicosmos86** [comment]
Hi @JoeSu112,

First, are you using GPU for inference? If you have switched from using CPU to GPU as the backed for JAX, then it might not be possible to use parallel inference. In the error messages that you have provided, it seems to be checking for CUDA versions, which led me to think that you might have installed the GPU version of JAX.

If you are unable to run different chains in parallel, then that could be the reason why estimation takes much longer. It might be helpful to paste all the packages and their versions in your virtual environment so we can start from there.

Thanks!
Paul

**Reply by JoeSu112** [reply]
Hi @digicosmos86,

Thanks so much for your reply. 
Here are the packages and their versions in my virtual environment:

```
Package                  Version
------------------------ ---------------------
absl-py                  2.1.0
arviz                    0.18.0
bambi                    0.13.0
blackjax                 1.1.1
boltons                  23.0.0
brotlipy                 0.7.0
cachetools               5.3.1
certifi                  2023.5.7
cffi                     1.15.1
cftime                   1.6.2
charset-normalizer       2.0.4
chex                     0.1.86
cloudpickle              2.2.1
coloredlogs              15.0.1
conda                    23.5.2
conda-content-trust      0.1.3
conda-libmamba-solver    23.5.0
conda-package-handling   2.1.0
conda_package_streaming  0.8.0
cons                     0.4.6
contourpy                1.1.0
cryptography             39.0.1
cycler                   0.11.0
Cython                   3.0.2
dill                     0.3.7
dm-tree                  0.1.8
et-xmlfile               1.1.0
etuples                  0.3.9
fastprogress             1.0.3
filelock                 3.12.3
flatbuffers              23.5.26
fonttools                4.42.1
formulae                 0.5.0
fsspec                   2023.9.0
graphviz                 0.20.1
h5netcdf                 1.3.0
h5py                     3.11.0
hddm-wfpt                0.1.1
HSSM                     0.2.1
huggingface-hub          0.15.1
humanfriendly            10.0
idna                     3.4
jax                      0.4.25
jaxlib                   0.4.25+cuda11.cudnn86
jaxopt                   0.8.3
joblib                   1.3.2
jsonpatch                1.32
jsonpointer              2.1
kiwisolver               1.4.5
libmambapy               1.4.1
logical-unification      0.4.6
markdown-it-py           3.0.0
matplotlib               3.7.2
mdurl                    0.1.2
miniKanren               1.0.3
ml-dtypes                0.2.0
mpmath                   1.3.0
multipledispatch         1.0.0
netCDF4                  1.6.4
numpy                    1.26.4
numpyro                  0.14.0
nvidia-cublas-cu11       11.11.3.6
nvidia-cuda-cupti-cu11   11.8.87
nvidia-cuda-nvcc-cu11    11.8.89
nvidia-cuda-runtime-cu11 11.8.89
nvidia-cudnn-cu11        9.1.0.70
nvidia-cufft-cu11        10.9.0.58
nvidia-cusolver-cu11     11.4.1.48
nvidia-cusparse-cu11     11.7.5.86
nvidia-nccl-cu11         2.21.5
onnx                     1.14.1
onnxruntime              1.15.1
openpyxl                 3.1.2
opt-einsum               3.3.0
optax                    0.2.2
packaging                23.0
pandas                   2.1.0
patsy                    0.5.3
Pillow                   10.0.0
pip                      24.0
pluggy                   1.0.0
protobuf                 4.24.3
psutil                   5.9.5
pycosat                  0.6.4
pycparser                2.21
Pygments                 2.17.2
pymc                     5.13.1
pymc3                    3.11.4
pyOpenSSL                23.0.0
pyparsing                3.0.9
PySocks                  1.7.1
pytensor                 2.20.0
python-dateutil          2.8.2
pytz                     2023.3.post1
PyYAML                   6.0.1
requests                 2.29.0
rich                     13.7.1
ruamel.yaml              0.17.21
scikit-learn             1.3.0
scipy                    1.10.1
seaborn                  0.13.2
semver                   3.0.2
setuptools               67.8.0
six                      1.16.0
ssm-simulators           0.7.2
statsmodels              0.14.0
sympy                    1.12
Theano                   1.0.5
Theano-PyMC              1.1.2
threadpoolctl            3.2.0
toolz                    0.12.0
tqdm                     4.65.0
typing_extensions        4.7.1
tzdata                   2023.3
urllib3                  1.26.16
wheel                    0.38.4
xarray                   2023.8.0
xarray-einstats          0.6.0
zstandard                0.19.0
```

Also, I wonder if there is any method to install the past version in order to provide you more information about the problems I've encountered as I remember that the error messages never showed up when I ran the estimation in the past.

Thanks so much again for your help!!

**Reply by digicosmos86** [reply]
It seems that you do have the `cuda` version of JAX installed (you have this line `jaxlib                   0.4.25+cuda11.cudnn86`). Is this intended?

Yes, since you have installed `hssm` through git to get the latest version, you can remove `hssm` with `pip uninstall hssm` and then install `hssm` from PyPI with `pip install hssm`.

**Reply by JoeSu112** [reply]
In the tutorial, "Sampling with JAX support for GPU", it says that:
```
pip install --upgrade "jax[cuda11_pip]<0.4.16" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```
However, the newest hssm suggests that jax version should >=0.4.25, < 0.5.0, so I change 0.4.16 to 0.4.26 then install.
And it became the version (also that line) you saw above.

That's good, I will install the old version to see the difference.

**Reply by digicosmos86** [reply]
Do you know the CUDA and cudnn versions on the machine where you are running HSSM? It could be that the CUDA versions are not compatible.

We did require JAX version to be lower than 0.4.16 to avoid some compatibility issues. However, this changed in 0.2.1. Now we require a much higher version of JAX to avoid compatibility issues.

**Reply by AlexanderFengler** [reply]
As a general note @JoeSu112, 

if you are using the gpu backend, from what I gather, parallelization across chains should happen across GPUs, which could be one issue here.

It is often still useful to run on GPU because speedups can be very large, but chain parallelization is more tricky. 

Did the runtimes you reported (4h --> 24h) actually pan out? I often see wildly inaccurate early time estimates during tuning. 

Best,
Alex

**Reply by JoeSu112** [reply]
Hi, @digicosmos86,
Sorry for the late reply. After trying for days, I find that GPU  is not activated at all. I ran the estimation on CPU all the time as I found that the following message had showed up when I ran estimation in the past version of HSSM:
```
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.) 
```
I will contact the maintainer of the machine to see how to activate CUDA on it.
Also, this is the information of CUDA on the machine:
```
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Thu_Nov_18_09:45:30_PST_2021
Cuda compilation tools, release 11.5, V11.5.119
Build cuda_11.5.r11.5/compiler.30672275_0
```
```
nvidia-smi      
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA RTX A6000               Off |   00000000:3B:00.0 Off |                  Off |
| 46%   72C    P2            144W /  300W |    2482MiB /  49140MiB |     67%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA RTX A6000               Off |   00000000:AF:00.0 Off |                  Off |
| 30%   39C    P8             23W /  300W |      12MiB /  49140MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+                                                                                         
```
But it's still strange that if the GPU is not activated all the time, why the estimation time become much longer in new version? The estimation time is indeed more than a day @AlexanderFengler. The estimation result is almost the same as I did in the past version.

Thanks so much again for your help!!

**Reply by digicosmos86** [reply]
Do you see this warning `WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.` when PyMC is imported?

**Reply by JoeSu112** [reply]
Yes, this warning shows up:
```python3
>>> import pytensor  # Graph-based tensor library
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
```

**Reply by digicosmos86** [reply, marked as answer]
Yeah this is the culprit. The problem is that newer numpy versions (>=1.26.4) stopped providing PyTensor with information about BLAS libraries, so newer versions of HSSM that depends on newer versions of PyMC (and Pytensor) will not know where to find the BLAS library.

There is an easy fix to this, which is to install PyMC in a conda environment before installing HSSM. We have a tutorial [here](https://github.com/lnccbrown/HSSM/blob/main/docs/getting_started/installation.md). So basically you need to create a conda enivornment, install JAX (the GPU version with cuda versions consistent with the cuda installed on the system) and PyMC with conda, and then install HSSM via pip. We are making lots of progress in making HSSM available through conda as well, so hopefully soon hssm can be installed directly with `conda install -c conda-forge hssm`.

**Comment by RaoXie** [comment]
> Hi all, I recently updated HSSM to the newest version (0.2.1). However, while I run the program that was able to run in the old version (0.2.0b1), some problem occurs:
> 
> ```python
> infer_DDM = DDM.sample(
>     sampler="nuts_numpyro",  # type of sampler to choose,
>     cores=2,  # how many cores to use
>     chains=chains,  # how many chains to run
>     draws=draws,  # number of draws from the markov chain
>     tune=int(draws*0.2),  # number of burn-in samples
>     idata_kwargs=dict(log_likelihood=False)
> ```
> 
> ```
> UserWarning: There are not enough devices to run parallel chains: expected 4 but got 2. 
> Chains will be drawn sequentially. If you are running MCMC in CPU, 
> consider using `numpyro.set_host_device_count(4)` at the beginning of your program. 
> You can double-check how many devices are available in your system using `jax.local_device_count()`.
> ```
> 
> I'm sure that I could run parallel chains in the past versions. So I'm wondering how I can solve this problem and why this did not happen in the past versions. Thanks for any help!!!

@JoeSu112 Hi, I am also trying to use slurm scheduler to run multiple chains in parallel:
![image](https://github.com/lnccbrown/HSSM/assets/93839122/99785c2d-e6e4-4f88-b62a-fc383695d355)
it worked in jupyter notebook, but for slurm it keeps giving me this error
![image](https://github.com/lnccbrown/HSSM/assets/93839122/66712751-4c21-4a4b-898a-5a48a1787467)
It seems has something to do with resources, I believe 1 node, 1 task, 4 cpu should be good for the slurm job. I also played around different combinations, for example 4 tasks, but still doesn't work. Do you mind sharing how did you run this through slurm? Thank you so much in advance for your help!
![image](https://github.com/lnccbrown/HSSM/assets/93839122/734832f7-2a64-4d98-aa57-66abb3dcc3f2)

Best,
Rao
